src/NLog/Targets/FileTarget.cs

```diff
@@ -2493,25 +2500,27 @@ private void WriteHeaderAndBom(BaseFileAppender appender)
             if (Header is null && !WriteBom) return;
 
             var length = appender.GetFileLength();
-            //  Write header and BOM only on empty files or if file info cannot be obtained.
-            if (length is null || length == 0)
-            {
-                if (WriteBom)
-                {
-                    InternalLogger.Trace("{0}: Write byte order mark from encoding={1}", this, Encoding);
-                    var preamble = Encoding.GetPreamble();
-                    if (preamble.Length > 0)
-                        appender.Write(preamble, 0, preamble.Length);
-                }
-
-                if (Header != null)
-                {
-                    InternalLogger.Trace("{0}: Write header", this);
-                    ArraySegment<byte> headerBytes = GetLayoutBytes(Header);
-                    if (headerBytes.Count > 0)
-                    {
-                        appender.Write(headerBytes.Array, headerBytes.Offset, headerBytes.Count);
-                    }
+            var isNewFile = length is null || length == 0;
+            var writeBom = isNewFile && WriteBom;
+            var writeHeader = Header != null && (isNewFile || WriteHeaderOnInitialFileOpen);
+            
+            //  Write BOM only on empty files or if file info cannot be obtained
+            if (writeBom)
+            {
+                InternalLogger.Trace("{0}: Write byte order mark from encoding={1}", this, Encoding);
+                var preamble = Encoding.GetPreamble();
+                if (preamble.Length > 0)
+                    appender.Write(preamble, 0, preamble.Length);
+            }
+          
+            //  Write Header only on empty files or if file info cannot be obtained or WriteHeaderOnInitialFileOpen is true
+            if (writeHeader)
```

Comment:
Can this become `if (Header != null && (isNewFile || WriteHeaderOnInitialFileOpen)`